Reworked importer CSV parsing

tinwelint opened the pull request:
For higher performance, higher parallelism and less garbage.

Instead of going through the stages w/ the queuing and passing batches,
importing is done by fully executing the importing threads in parallel.
All data structures and access patterns have been made to allow for
concurrent access, mostly by sharing nothing.

Garbage was a big concern previously and so this new importing doesn't
create objects for the entities, but instead allow visitors to observe
the different fields, directly from an underlying buffer which has been
read from the input source.

tinwelint committed:
Rewrite of how data from Input gets written into records

Instead of going through the stages w/ the queuing and passing objects
importing is done by fully executing the importing threads in parallel.
All data structures and access patterns ihave been made to allow for
concurrent access.

Garbage was a big concern previously and so this new importing doesn't
create objects for the entities, but instead allow visitors to observe
the different fields, directly from an underlying buffer which has been
read from the input source.

tinwelint committed:
Recycles and reuses batch objects for way less garbage

tinwelint committed:
Fixes a concurrency record counting issue

tinwelint committed:
Reinstates support for multiline-fields

tinwelint committed:
Fixes a bunch of test problems

tinwelint committed:
Small fixes and added javadoc

tinwelint committed:
Fixes some issues with migration case of the importer

tinwelint committed:
Fixes label index deletion issue deduping nodes

Since label index is built in the end there's no need for including it when
deleting duplicate nodes.

tinwelint commented:
Changes in this branch which makes CSV --> record ids non-deterministic may break algorithms in development for future features, so putting this on ice a bit.

tinwelint committed:
Updated license headers for 2018

tinwelint committed:
Removed unused class

tinwelint committed:
Displays correct import counts

tinwelint committed:
Importer correctly deletes everything about duplicate nodes

Even dynamic label records and property records

tinwelint committed:
Smaller group cache in EncodingIdMapper if possible

tinwelint committed:
Fixes a relationship counting issue in DataStatistics

tinwelint committed:
Always notifies Collector about bad relationships

so that exceptions propagate as expected

tinwelint committed:
Import test uses one file to avoid race

Race would be that there would be two parallel processors each with data
from different files and the one that happened to report the bad relationship
first would win and produce the exception message.

tinwelint committed:
Improves a silly javadoc.

chrisvest commented on community/import-tool/src/test/java/org/neo4j/tooling/ImportToolTest.java:
```diff
@@ -1281,6 +1315,7 @@ private void shouldPrintReferenceLinkAsPartOfErrorMessage( List<String> nodeIds,
         }
     }
 
+    @Ignore
```
Should this still be ignored?

chrisvest replied:
The test seems to pass.

chrisvest commented on community/import-tool/src/test/java/org/neo4j/tooling/ImportToolTest.java:
```diff
@@ -1566,6 +1603,7 @@ public void shouldBeEquivalentToUseRawAsciiOrCharacterAsQuoteConfiguration1() th
         }
     }
 
+    @Ignore
```
This test is not passing, but maybe it just need to be updated to the new error format?

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/BatchingIdGetter.java:
```diff
@@ -0,0 +1,72 @@
+ * It makes use of {@link IdSequence#nextIdBatch(int)} (with default batch size the number of records per page)
+ * and caches that batch, exhausting it in {@link #next()} before getting next batch.
+ *
+ * TODO use the {@link RenewableBatchIdSequences} instead.
```
TODO?

tinwelint replied:
Removed TODO note, not really applicable.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/DataStatistics.java:
```diff
@@ -19,56 +19,184 @@
+        {
+            if ( typeId >= counts.length )
+            {
+                counts = Arrays.copyOf( counts, max( counts.length * 2, typeId ) );
```
This needs to be `typeId + 1`, since the parameter here is the _length_, while `typeId` is elsewhere used as an _index_.

chrisvest replied:
And this should be tested, in case there are other places in the code that aren't fully aligned on this.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/DataImporter.java:
```diff
@@ -0,0 +1,268 @@
+ */
+public class DataImporter
+{
+    public static final String ID_PROPERTY = "__id";
```
`ID_PROPERTY` is never used.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/ImportLogic.java:
```diff
@@ -150,18 +141,13 @@ public void initialize( Input input ) throws IOException
-        cachedNodes = cachedForSure( nodes, inputCache.nodes( MAIN, true ) );
         Estimates inputEstimates = input.calculateEstimates( neoStore.getPropertyStore().newValueEncodedSizeCalculator() );
         sanityCheckEstimatesWithRecordFormat( inputEstimates );
         dependencies.satisfyDependencies( inputEstimates, idMapper, neoStore, nodeRelationshipCache );
```
There are some `System.out` and `System.err` print-outs in this class. What's that about?

tinwelint replied:
The import tool generally prints things to `out`, `err`, although this is a bit down the stack so it could be done using a monitor instead I'd guess.

chrisvest replied:
At least the PrintStreams should be passed in, instead of referencing `System.out/err` directly.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/EntityImporter.java:
```diff
@@ -0,0 +1,173 @@
+
+    private void encodeProperty( PropertyBlock block, int key, Object value )
+    {
+        // TODO: dynamic record ids, batching of those
```
TODO?

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/DataImporter.java:
```diff
@@ -0,0 +1,268 @@
+        ControllableStep step = new ControllableStep( title, roughEntityCountProgress, Configuration.DEFAULT,
+                writeMonitor, memoryStatsProvider );
+        StageExecution execution = new StageExecution( title, null, Configuration.DEFAULT, Collections.singletonList( step ), 0 );
+        InputIterator dataIterator = data.iterator();
```
The `dataIterator` will be closed by each of the `ExhaustingEntityImporterRunnable`s. So it will be closed `numRunners` times. I think we should instead close it at this level, with a `try-with-resources` that spans until the executor has been shut down.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/DataImporter.java:
```diff
@@ -0,0 +1,268 @@
+            throws IOException
+    {
+        importData( NODE_IMPORT_NAME, numRunners, input.nodes(), stores, () ->
+            new NodeImporter( stores, idMapper, monitor ), executionMonitor, new MemoryUsageStatsProvider( stores, idMapper ) );
```
How about having methods for creating these lambdas? The `NodeImporter` one is not so bad, but the `RelationshipImporter` involves a lot of fields.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/LabelIndexWriterStep.java:
```diff
@@ -62,7 +60,7 @@ protected void process( Batch<InputNode,NodeRecord> batch, BatchSender sender )
     {
-        super.close();
         writer.close();
+        super.close();
```
The `super.close()` method will close the executor, after which the task queues will be empty and all workers will have terminated. If we close the `writer` before we close the executor, we can risk racing with on-going work, can we not?

tinwelint replied:
In a happy scenario (non-panic) this will not be a problem since `close()` will be called after all work has completed. I don't mind changing this order anyway.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/EncodingIdMapper.java:
```diff
@@ -748,15 +682,15 @@ private long dataValue( long index )
         // can't be done on unsorted data
-        long low = 0;
-        long high = highestSetIndex;
+        long low = 0 + 0;
```
Just `0` will do.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/EncodingIdMapper.java:
```diff
@@ -37,26 +39,28 @@
 
 /**
  * Maps arbitrary values to long ids. The values can be {@link #put(Object, long, Group) added} in any order,
  * but {@link #needsPreparation() needs} {@link #prepare(InputIterable, Collector, ProgressListener) preparation}
```
This part of the javadocs needs to be updated as well.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/EncodingIdMapper.java:
```diff
@@ -344,13 +315,15 @@ private static boolean isCollision( long eId )
 
         private int numberOfCollisions;
         private int localProgress;
+        private final LongFunction<Object> inputIdLookup;
```
This field is not used.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/EncodingIdMapper.java:
```diff
@@ -609,60 +567,50 @@ public long dataValue( long nodeId )
+            // number of collisions. But that's probably alright since the data structures and
+            // actual collisions values for all these collisions wouldn't fit in a heap anyway.
+            Object inputId = collisionValues.get( safeCastLongToInt( collisionIndex ) );
+            int detectorIndex = detector.add( inputId );
```
The `add` method loops through its `inputIdArray`, and we are calling it in a loop over the collisions. So we have a quadratic algorithm here. How likely is that to be a problem?

tinwelint replied:
In normal cases there are, say 0 to a hand full collisions for some of the values. I wouldn't consider that a problem for 99.99% of all use cases.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/StringEncoder.java:
```diff
@@ -127,7 +127,13 @@ private void reMap( byte[] bytes, int inputLength )
+                    }
+                }
             }
             bytes[i] = reMap[bytes[i]];
```
This double-check locking will not work, because the `reMap[bytes[i]]` loads are not volatile. There is also no coordination between the load in the if-condition, and the load that assigns to `bytes[i]`. The value at that array index could be changed by another thread in between the two loads.

Documentation and testing of this new multi-threaded behaviour is also missing.

tinwelint replied:
It's somewhat unrelated to this PR though, I can create a card for it

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/Tracker.java:
```diff
@@ -30,7 +30,15 @@
+=======
+ * {@link IdMapper#prepare(java.util.function.LongFunction, Collector, ProgressListener) prepare phase} where these ids are sorted
+>>>>>>> Rewrite of how data from Input gets written into records
+>>>>>>> c82c373... Rewrite of how data from Input gets written into records
```
Your merge conflicts are showing.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/ByteBufferReadableChannel.java:
```diff
@@ -0,0 +1,85 @@
+/**
+ * TODO a bit overkill to wrap a byte[] -> {@link ByteBuffer} -> {@link ReadableClosableChannel}?
+ */
+public class ByteBufferReadableChannel implements ReadableClosableChannel
```
Judging by usages, it seems that ValueType is working (by proxy) on ByteBuffers anyway, rather than real channels. Maybe that should be surfaced in the parameter lists to the read and write methods on ValueType? Then you wouldn't need these classes, but you'd have to fix that (as far as I have found) one place where a "real" channel is being worked on.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/Groups.java:
```diff
@@ -28,9 +28,11 @@
  */
 public class Groups
 {
+    public static final int MAX_NUMBER_OF_GROUPS = 0x10000;
```
`MAX_NUMBER_OF_GROUPS` isn't used anywhere.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputCache.java:
```diff
@@ -250,4 +233,20 @@ public void close() throws IOException
+        return ByteBuffer.allocate( Integer.BYTES );
+    }
+
+    static String sample( ByteBuffer buffer )
```
The `sample` method is debugging leftovers?

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputCache.java:
```diff
@@ -128,47 +128,33 @@
     private final FileSystemAbstraction fs;
     private final File cacheDirectory;
     private final RecordFormats recordFormats;
-    private final Configuration config;
```
The `NO_ENTITIES` and `END_OF_CACHE` constants are no longer used.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputEntityCacher.java:
```diff
@@ -44,212 +41,263 @@
+                buffer.position( position );
+                bufferAsChannel = new ByteBufferFlushableChannel( buffer );
+            }
+            return buffer;
```
This will work because the chunk size is so large to begin with, and this method is only called with small constants. But there is bug-potential here, where `position + withSufficientSpaceFor` could in principle be greater than twice the buffer capacity. Also, `withSufficientSpaceFor` could be called `requiredSpace`, which is shorter.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputEntityCacher.java:
```diff
@@ -44,212 +41,263 @@
+            catch ( Exception e )
+            {
+                throw e;
+            }
```
Catch and immediately re-throw?

tinwelint replied:
Some old remnant I believe.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputEntityCacher.java:
```diff
@@ -44,212 +41,263 @@
+        private String stringify( Object value )
+        {
+            return value.getClass().isArray() ? Arrays.toString( (Object[]) value ) : value.toString();
+        }
```
Debugging leftovers?

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputEntityCacher.java:
```diff
@@ -44,212 +41,263 @@
+ * Currently each token type is limited to have a maximum of {#link Integer.MAX_VALUE} items.
  */
-abstract class InputEntityCacher<ENTITY extends InputEntity> implements Receiver<ENTITY[],IOException>
+abstract class InputEntityCacher implements InputCacher
```
Please add more documentation of what is going on here, and how it's thread-safe. This class is hard to follow.

chrisvest replied:
If possible, I'd recommend separating domain logic and multi-threading concerns in the code. Things are always more complicated when the two are entangled.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputEntity.java:
```diff
@@ -19,204 +19,291 @@
-        int collidingKeys = 0;
-        for ( int i = 0; i < properties.length; i += 2 )
+        checkClear();
+        for ( String label : labels )
```
`Collections.addAll`

tinwelint replied:
And create a new object? Perhaps it'll not be an actual instance after escape analysis, or?

chrisvest replied:
`Collections.addAll` is a static method that iterates an array and adds all the array elements to a given collection. The same as what this code does.

tinwelint replied:
Cheers, I'll use that one

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputEntity.java:
```diff
@@ -19,204 +19,291 @@
+    @Override
+    public InputEntity clone() throws CloneNotSupportedException
+    {
+        InputEntity clone = new InputEntity();
```
The clone instance should always be acquired from calling `super.clone()`. Once you have that, you only need replace the array list instances with their clones in turn. Take a look at the `Object.clone()` javadocs.

tinwelint replied:
Cool, didn't know that.

tinwelint replied:
Although that will force those lists to be non-final, won't it?

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputEntity.java:
```diff
@@ -19,204 +19,291 @@
+        clone.hasIntType = hasIntType;
+        clone.intType = intType;
+        clone.stringType = stringType;
+        return clone;
```
This implementation of `clone` does not retain the delegate field. Is that intended?

tinwelint replied:
Hmm, don't think it's used actually. The clone() method that is.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputEntityCacher.java:
```diff
@@ -44,212 +41,263 @@
-            channel.put( SAME_GROUP );
+            super( actual );
+            this.lengthThreshold = chunkSize;
+            this.array = new byte[chunkSize + chunkSize / 10]; // some wiggle room
```
Same as `chunkSize / 5`?

tinwelint replied:
More like `chunkSize` plus `chunkSize/10`

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputEntityReader.java:
```diff
@@ -161,159 +124,114 @@ private void readHeader( StoreChannel header ) throws IOException
-
-            return readNextOrNull( properties, state );
+            channel.close();
+            closeAction.run();
```
Hmm… The `StandardOpenOption.DELETE_ON_CLOSE` would have been a perfect alternative, had our `FileSystemAbstraction` supported them.

chrisvest replied:
Except for the maintenance of that `subtype` set, whatever that is.

tinwelint replied:
Turns out that this is old unused stuff anyway

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputEntityReader.java:
```diff
@@ -161,159 +124,114 @@ private void readHeader( StoreChannel header ) throws IOException
+                    Object token = readToken( PROPERTY_KEY_TOKEN );
+                    Object value = readValue( );
+                    // TODO fix this some other way
+                    if ( token instanceof Integer )
```
What are the plans for this TODO?

tinwelint replied:
None, really.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputEntityReader.java:
```diff
@@ -161,159 +124,114 @@ private void readHeader( StoreChannel header ) throws IOException
+                visitor.propertyId( channel.getLong() );
+                break;
+            case 0:
+                break;
```
The `default` case seems to handle `0` just as well. That brings us down to 3 cases, at which point a switch-statement might be overkill on the ceremony.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputEntityReader.java:
```diff
@@ -161,159 +124,114 @@ private void readHeader( StoreChannel header ) throws IOException
+            {
+                // This is a real token id
+                return channel.getInt();
+            }
```
When does this happen?

tinwelint replied:
When the cached input already contains real token ids, they are allowed to.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputEntityReader.java:
```diff
@@ -51,104 +40,78 @@
  * the parallelization happens inside.
  */
-abstract class InputEntityReader<ENTITY extends InputEntity> extends InputIterator.Adapter<ENTITY>
+abstract class InputEntityReader implements InputIterator
```
This class could be named `InputEntityCacheReader`, and the `InputEntityCacher` could be called `InputEntityCacheWriter`, to indicate that they are siblings around this cache format.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputChunk.java:
```diff
@@ -0,0 +1,297 @@
+        {
+            return false;
+        }
+        return true;
```
The return could be simplified to the expression.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/csv/Decorator.java:
```diff
@@ -22,9 +22,9 @@
+import org.neo4j.unsafe.impl.batchimport.input.InputEntityVisitor;
 
-public interface Decorator<ENTITY extends InputEntity> extends Function<ENTITY,ENTITY>, Resource
+public interface Decorator extends Function<InputEntityVisitor,InputEntityVisitor>, Resource
```
This implements `Resource` to get a `close` method, which we then provide a default empty implementation of. But we don't actually use this feature anywhere. Let's just not make it extend `Resource`.

tinwelint replied:
Again this is for an external use case in the ldbc importer which I'm not sure we can live without.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/csv/Header.java:
```diff
@@ -37,8 +39,12 @@
-        Header create( CharSeeker dataSeeker, Configuration configuration, IdType idType );
+        Header create( CharSeeker dataSeeker, Configuration configuration, IdType idType, Groups groups );
+
+        boolean isDefined();
```
What does this mean?

tinwelint replied:
Added javadoc

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/ByteBufferFlushableChannel.java:
```diff
@@ -27,7 +27,7 @@
 
 /**
- * TODO a bit overkill to wrap a byte[] -> {@link ByteBuffer} -> {@link ReadableClosableChannel}?
+ * Wraps a byte[] -> {@link ByteBuffer} -> {@link ReadableClosableChannel}?
```
These two still have question marks at the end.

chrisvest commented on community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/RandomEntityDataGenerator.java:
```diff
@@ -0,0 +1,221 @@
+    private static Object randomProperty( Entry entry, Randoms random )
+    {
+        // TODO crude way of determining value type
+        String type = entry.extractor().name();
```
Yeah, perhaps an enum would be better.

tinwelint replied:
Dunno if it's worth adding that, since it's only test code caring about it a.t.m. OK to just remove the TODO?

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputIterator.java:
```diff
@@ -0,0 +1,135 @@
+    static Header extractHeader( CharReadable stream, Header.Factory headerFactory, IdType idType,
+            Configuration config, Groups groups ) throws IOException
+    {
+        if ( !headerFactory.isDefined() )
```
Looks like this is only ever `false` in testing. Is it necessary to have this?

tinwelint replied:
I do think that there's one external use of this in the ldbc importer as well tho.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/staging/RecordDataAssembler.java:
```diff
@@ -0,0 +1,78 @@
+        {
+            return false;
+        }
+        return true;
```
This can be simplified to return an expression. And the null-check could be avoided by initialising the `filter` field to `Predicates.alwaysTrue` if no filter is otherwise given in the constructor.

chrisvest commented on community/csv/src/main/java/org/neo4j/csv/reader/Chunker.java:
```diff
@@ -0,0 +1,57 @@
+     * to fill it with data. When a {@link Chunk} has been fully processed then it can be passed into
+     * {@link #nextChunk(Chunk)} again to get more data.
+     */
+    Chunk newChunk();
```
Looks like `newChunk` is only really used in tests, and `EagerParserChunker` doesn't support it anyway?

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputIterator.java:
```diff
@@ -0,0 +1,135 @@
+            return chunker.nextChunk( csvChunk );
+        }
+
+        CsvInputChunk csvChunk = (CsvInputChunk) chunk;
```
These casts are quite smelly. They seem to indicate that this chunk abstraction doesn't really work here.

tinwelint replied:
This will be slightly rewritten, but as with an API which has explicit allocation like:

````
MyInterface
{
    MyThing allocateAThing();

    void fillMyThingWithSomething( MyThing thing );
}
````

This cast comes by design, doesn't it?

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/csv/EagerlyReadInputChunk.java:
```diff
@@ -0,0 +1,91 @@
+    @Override
+    public char[] data()
+    {
+        throw new UnsupportedOperationException();
```
All the methods from `Source.Chunk` are unsupported. Why is it implementing this interface? 🤔

tinwelint replied:
It's this weird thing where there's the normal lazy chunk/parser and then we have to, against our will, support the multi-line fields feature and it's conflation of newline with record delimiters... this changes how parsing can be done significantly, for that case only... but it needs to fit in with the rest at the same time.

I'd guess this is a turd from that behind. I'm sure it can be done better, I just don't see it clearly a.t.m.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputEntity.java:
```diff
@@ -19,204 +19,373 @@
+
+    public Object endId()
+    {
+        return hasLongEndId ? longEndId : objectEndId;
```
Could the auto-box be avoided here? Or will exposing getters for each of the `has*`, `long*` and `object*` fields just move the same auto-boxing further up the stack?

tinwelint replied:
I think it would unfortunately. This isn't part of the main paths anyway.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/ByteBufferReadableChannel.java:
```diff
@@ -0,0 +1,85 @@
+/**
+ * Wraps a byte[] -> {@link ByteBuffer} -> {@link ReadableClosableChannel}?
+ */
+public class ByteBufferReadableChannel implements ReadableClosableChannel
```
Judging by usages, it seems that ValueType is working (by proxy) on ByteBuffers anyway, rather than real channels. Maybe that should be surfaced in the parameter lists to the read and write methods on ValueType? Then you wouldn't need these classes, but you'd have to fix that (as far as I have found) one place where a "real" channel is being worked on.

tinwelint replied:
I know I struggled a bit with that back and forth... I'd be happy to let it rest, but the wrapping is a bit annoying, even though it's just a very peripheral thing in the outer corner of the import code.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/DataImporter.java:
```diff
@@ -0,0 +1,291 @@
+                    throws IOException
+    {
+        importData( NODE_IMPORT_NAME, numRunners, input.nodes(), stores, () ->
+            new NodeImporter( stores, idMapper, monitor ), executionMonitor, new MemoryUsageStatsProvider( stores, idMapper ) );
```
How about having methods for creating these lambdas? The NodeImporter one is not so bad, but the RelationshipImporter involves a lot of fields.

tinwelint replied:
You mean introduce variable? Sure

chrisvest replied:
Variable or method. Either works.

chrisvest reviewed (changes requested):
I had a couple of comments. Please address them in _new_ commits.

ragadeeshu committed:
Changed CsvGroupInputIterator <--> InputChunk relationship

By pushing down implementation details about the different types of
chunks (eager vs lazy) further down.

tinwelint committed:
Unignored some tests in ImportToolTest

tinwelint committed:
Batches id generator of dynamic property records

tinwelint committed:
Better lifecycle and more readable DataImporter

tinwelint committed:
Fixes an issue growing type count array in importer

tinwelint committed:
Moves some printouts to a monitor

tinwelint committed:
Changes order of closing resources

tinwelint committed:
Some cleanup in EncodingIdMapper

tinwelint committed:
Cleanup some doc and unused code

tinwelint committed:
Cleanup of InputEntityCacher

tinwelint committed:
Renamed InputEntityCacher and InputEntityReader

To InputEntityCacheWriter and InputEntityCacheReader respectively, also
its subclasses accordingly.

tinwelint committed:
Removes old remnants of subTypes in InputCache

tinwelint committed:
Cleanup of InputEntityCacheReader

tinwelint committed:
Some clarifications and simplifications

tinwelint commented:
@chrisvest thanks, addressed most comments

tinwelint committed:
Uses utility method for adding labels
